packages/daemon/src/service/trace-manager.ts

```diff
@@ -0,0 +1,251 @@
+ */
+export class TraceEvent {
+  data?: any;
+  constructor(public type: TraceType) {}
+}
+
+/**
+ * pipcook event data type for job status change
+ */
+export class JobStatusChangeEvent extends TraceEvent {
+  data: {
+    jobStatus: PipelineStatus;
+    step?: PluginTypeI;
+    stepAction?: 'start' | 'end';
+    queueLength?: number;
+  };
+  constructor(jobStatus: PipelineStatus, step?: PluginTypeI, stepAction?: 'start' | 'end', queueLength?: number) {
+    super('job_status');
+    this.data = { jobStatus, step, stepAction, queueLength };
+  }
+}
+
+type LogLevel = 'info' | 'warn' | 'error';
+/**
+ * pipcook event data type for log
+ */
+export class LogEvent extends TraceEvent {
+  data: {
+    level: LogLevel;
+    data: string;
+  };
+  constructor(level: LogLevel, data: string) {
+    super('log');
+    this.data = { level, data };
+  }
+}
+
+export type PipcookEvent = JobStatusChangeEvent | LogEvent;
+/**
+ * trace handler
+ * it has 2 parts: logger and event handler:
+ * logger:
+ * stdout and stderr, they are streams to pipe the logs to clients
+ * event handler:
+ * pipe the pipcook event to clients
+ */
+export class Tracer {
+  // trace id
+  id: string;
+  // stdout stream for log pipe
+  private stdout: LogPassthrough;
+  // stderr stream for log pipe
+  private stderr: LogPassthrough;
+  // event emitter for pipcook event
+  private emitter: EventEmitter;
+  // options of tracer
+  private opts?: TraceOptions;
+  // log file fd for stdout, -1 if not defined
+  private fdOut: number;
+  // log file fd for stderr, -1 if not defined
+  private fdErr: number;
+
+  constructor(opts?: TraceOptions) {
+    this.id = generateId();
+    this.emitter = new EventEmitter();
+    this.opts = opts;
+  }
+
+  /**
+   * init loggers
+   */
+  async initLogger() {
+    this.fdOut = this.opts?.stdoutFile ? await open(this.opts?.stdoutFile, 'w+') : -1;
+    this.fdErr = this.opts?.stderrFile ? await open(this.opts?.stderrFile, 'w+') : -1;
+    this.stdout = new LogPassthrough(this.fdOut);
+    this.stderr = new LogPassthrough(this.fdErr);
+  }
+  /**
+   * get the loggers
+   */
+  getLogger(): { stdout: LogPassthrough; stderr: LogPassthrough } {
+    return { stdout: this.stdout, stderr: this.stderr };
+  }
+
+  /**
+   * listen event
+   * @param cb event callback
+   */
+  listen(cb: (data: PipcookEvent) => void): void {
+    // event callback
+    this.emitter.on('trace-event', (e) => {
+      cb(e);
+    });
+
+    // log callback
+    const pipeLog = (level: LogLevel, logger: LogPassthrough) => {
+      logger.on('data', data => {
+        cb(new LogEvent(level, data));
+      });
+      logger.on('close', () => this.emitter.emit('trace-finished'));
+      logger.on('error', err => {
+        cb(new LogEvent('error', err.message));
+      });
+    };
+    pipeLog('info', this.stdout);
+    pipeLog('warn', this.stderr);
+  }
+
+  /**
+   * emit event to client
+   * @param event pipcook event data
+   */
+  emit(event: PipcookEvent) {
```

Comment:
We may use `dispatchEvent(event)` instead of `emit`, see https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent.